js/components/authoring.js

```diff
@@ -17,104 +17,129 @@ const AVAILABLE_PRESETS = [
-        this.setState({step: 4})
-      } else {
+    if (step !== prevState.step) {
+      const stepName = this.currentStep
+      if (stepName === 'presets') {
+        this.unloadModel()
+      } else if (stepName === 'continents') {
+        this.setContinentsStep()
+      } else if (stepName === 'forces') {
+        this.setForcesStep()
+      } else if (stepName === 'densities') {
+        this.setDensitiesStep()
+      } else if (stepName === undefined) {
         this.endAuthoring()
       }
-    } else if (step === 4) {
-      this.endAuthoring()
+
+      if (step > prevState.step) {
+        this.saveModel()
+      } else if (stepName !== 'presets') {
+        this.restoreModel()
+      }
     }
   }
 
+  handleNextButtonClick () {
+    const { step } = this.state
+    this.setState({ step: step + 1 })
+  }
+
   handleBackButtonClick () {
     const { step } = this.state
-    const { restoreLabeledSnapshot } = this.props
-    if (step === 2) {
-      this.unloadModel()
-    } else if (step === 3) {
-      restoreLabeledSnapshot(STEP_NAMES[2])
-      this.setStep2()
-    } else if (step === 4) {
-      restoreLabeledSnapshot(STEP_NAMES[3])
-      this.setStep3()
+    if (step > 0) {
+      this.setState({ step: step - 1 })
     }
   }
 
-  handleInteractionChange (interactionName) {
-    const { setOption } = this.props
-    setOption('interaction', interactionName)
+  saveModel () {
+    const { takeLabeledSnapshot } = this.props
+    takeLabeledSnapshot(this.currentStep)
   }
 
-  snapshotStep () {
-    const { step } = this.state
-    const { takeLabeledSnapshot } = this.props
-    takeLabeledSnapshot(STEP_NAMES[step])
+  restoreModel () {
+    const { restoreLabeledSnapshot } = this.props
+    restoreLabeledSnapshot(this.currentStep)
   }
 
   loadModel (presetInfo) {
     const { loadModel } = this.props
-    loadModel(presetInfo.name)
-    this.setStep2()
+    this.setState({ modelLoading: true })
+    loadModel(presetInfo.name, () => {
+      // Important! Go to the next step after model load is complete.
+      // Otherwise, we might try to take model snapshot too early (when it's not loaded yet).
+      this.setState({ modelLoading: false })
+      this.handleNextButtonClick()
+    })
   }
 
   unloadModel () {
     const { unloadModel, setOption } = this.props
     unloadModel()
     setOption('interaction', 'none')
     setOption('selectableInteractions', [])
-    this.setState({step: 1})
   }
 
-  setStep2 () {
+  setContinentsStep () {
     const { setOption } = this.props
     setOption('interaction', 'drawContinent')
     setOption('selectableInteractions', [ 'drawContinent', 'eraseContinent', 'none' ])
-    this.setState({step: 2})
   }
 
-  setStep3 () {
+  setForcesStep () {
     const { setOption } = this.props
     setOption('interaction', 'force')
     setOption('selectableInteractions', [ 'force', 'none' ])
     setOption('colormap', 'topo')
-    this.setState({step: 3})
+  }
+
+  setDensitiesStep () {
```

Comment:
Again, we could put these in the `STEP` object if we wanted, and further clean up `componentDidUpdate`. that's what I would probably do if this was Java, but I'm not sure what best practices are for js

Reply:
Right, we could do that. It would be more "meta". I've looked at it, but there are still functions that have a bunch of "setOption" calls and can't be moved to STEP_DATA, e.g. initial component setup or setup after the wizard is finished. Also, some steps do more work, e.g. presets step unloads the model. So, I've finally left it as it is. But it's an option for sure if number of steps grows or we want to do more refactoring.